playback/src/local_file.rs

```diff
@@ -0,0 +1,182 @@
+use librespot_core::{Error, SpotifyUri};
+use regex::{Captures, Regex};
+use std::sync::LazyLock;
+use std::{
+    collections::HashMap,
+    fs,
+    fs::File,
+    io,
+    path::{Path, PathBuf},
+    time::Duration,
+};
+use symphonia::{
+    core::formats::FormatOptions,
+    core::io::MediaSourceStream,
+    core::meta::{MetadataOptions, StandardTagKey, Tag},
+    core::probe::{Hint, ProbeResult},
+};
+
+// "Spotify supports .mp3, .mp4, and .m4p files. It doesn’t support .mp4 files that contain video,
+// or the iTunes lossless format (M4A)."
+// https://community.spotify.com/t5/FAQs/Local-Files/ta-p/5186118
+//
+// There are some indications online that FLAC is supported, so check for this as well.
+const SUPPORTED_FILE_EXTENSIONS: &[&str; 4] = &["mp3", "mp4", "m4p", "flac"];
+
+#[derive(Default)]
+pub struct LocalFileLookup(HashMap<SpotifyUri, PathBuf>);
+
+impl LocalFileLookup {
+    pub fn get(&self, uri: &SpotifyUri) -> Option<&Path> {
+        self.0.get(uri).map(|p| p.as_path())
+    }
+}
+
+pub fn create_local_file_lookup(directories: &[PathBuf]) -> LocalFileLookup {
+    let mut lookup = LocalFileLookup(HashMap::new());
+
+    for path in directories {
+        if !path.is_dir() {
+            warn!(
+                "Ignoring local file source {}: not a directory",
+                path.display()
+            );
+            continue;
+        }
+
+        if let Err(e) = visit_dir(path, &mut lookup) {
+            warn!(
+                "Failed to load entries from local file source {}: {}",
+                path.display(),
+                e
+            );
+        }
+    }
+
+    lookup
+}
+
+fn visit_dir(dir: &Path, accumulator: &mut LocalFileLookup) -> io::Result<()> {
+    for entry in fs::read_dir(dir)? {
+        let path = entry?.path();
+        if path.is_dir() {
+            visit_dir(&path, accumulator)?;
+        } else {
+            let Some(extension) = path.extension().and_then(|e| e.to_str()) else {
+                continue;
+            };
+
+            let lowercase_extension = extension.to_lowercase();
+
+            if SUPPORTED_FILE_EXTENSIONS.contains(&lowercase_extension.as_str()) {
+                let uri = match get_uri_from_file(path.as_path(), extension) {
+                    Ok(uri) => uri,
+                    Err(e) => {
+                        warn!(
+                            "Failed to determine URI of local file {}: {}",
+                            path.display(),
+                            e
+                        );
+                        continue;
+                    }
+                };
+
+                accumulator.0.insert(uri, path);
+            }
+        }
+    }
+
+    Ok(())
+}
+
+fn get_uri_from_file(audio_path: &Path, extension: &str) -> Result<SpotifyUri, Error> {
+    let src = File::open(audio_path)?;
```

Comment:
All of my I/O for this feature is synchronous for now.

I tried using `tokio::fs` APIs but they don't appear to play ball with Symphonia's `MediaSource` trait bound, however maybe `into_std()` which I have just found while writing this comment could help.

Additionally, we create this lookup inside `player::new()`, I get that in Rust there are no constructors and `new()` is not special, but something about my C# background makes me very hesitant to make `new()` async, am I right to think this?

Reply:
you can make any function/method async, none has a specific meaning (as far as I'm aware of). But I can understand why you would prevent to make the initializer async.

Reply:
Yeah, whichever language I'm working in I usually try and make initialisers/constructors quite simple as I think that's what people expect (although technically I'm not following this by doing I/O in `new()`, that said `new()` was already spawning a thread so maybe it's not much of a change)

Additionally, making `Player::new` async would be quite a big breaking change. I think if we did want the initial lookup creation to be async we would want to instead pass in a `LocalFileLookup` directly into the player and expect users to create it themselves -- this is a bit harder to use as an API though so let me know what you think

Reply:
I mean keeping it sync as-is is probably fine. Just making it async of the sake of being async might be a bit overkill. To have the option would be nice, but it's probably something that's outside of this PR's scope.

Reply:
True, also I don't think symphonia (which is doing most of the heavy lifting here) has any async support, I think I read some comment to the effect of this part of the code is 'fake async' by running on a background thread because seeking was synchronous. I could make the `File::open` call async perhaps in case the symphonia limitation gets lifted later?